src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+            set
+            {
+                // Unlike OutputFilePath and OutputRefFilePath, the intermediate output path isn't represented in the workspace anywhere;
+                // this, we won't mutate the solution. We'll still call ChangeProjectOutputPath so we have the rest of the output path tracking
+                // for any P2P reference conversion.
+                ChangeProjectOutputPath(ref _intermediateOutputFilePath, value, s => s, w => { });
+            }
+        }
+
+        public string OutputFilePath
+        {
+            get => _outputFilePath;
+            set => ChangeProjectOutputPath(ref _outputFilePath,
+                       value,
+                       s => s.WithProjectOutputFilePath(Id, value),
+                       w => w.OnOutputFilePathChanged(Id, value));
+        }
+
+        public string OutputRefFilePath
+        {
+            get => _outputRefFilePath;
+            set => ChangeProjectOutputPath(ref _outputRefFilePath,
+                       value,
+                       s => s.WithProjectOutputRefFilePath(Id, value),
+                       w => w.OnOutputRefFilePathChanged(Id, value));
+        }
+
+        public string FilePath
+        {
+            get => _filePath;
+            set => ChangeProjectProperty(ref _filePath,
+                       value,
+                       s => s.WithProjectFilePath(Id, value),
+                       w => w.OnProjectNameChanged(Id, _displayName, value));
+        }
+
+        public string DisplayName
+        {
+            get => _displayName;
+            set => ChangeProjectProperty(ref _displayName,
+                       value,
+                       s => s.WithProjectName(Id, value),
+                       w => w.OnProjectNameChanged(Id, value, _filePath));
+        }
+
+        // internal to match the visibility of the Workspace-level API -- this is something
+        // we use but we haven't made officially public yet.
+        internal bool HasAllInformation
+        {
+            get => _hasAllInformation;
+            set => ChangeProjectProperty(ref _hasAllInformation,
+                       value,
+                       s => s.WithHasAllInformation(Id, value),
+                       w => w.OnHasAllInformationChanged(Id, value));
+        }
+
+        public string Language { get; }
+
+        #region Batching
+
+        public BatchScope CreateBatchScope()
+        {
+            lock (_gate)
+            {
+                _activeBatchScopes++;
+                return new BatchScope(this);
+            }
+        }
+
+        public sealed class BatchScope : IDisposable
+        {
+            private readonly VisualStudioProject _project;
+
+            /// <summary>
+            /// Flag to control if this has already been disposed. Not a boolean only so it can be used with Interlocked.CompareExchange.
+            /// </summary>
+            private volatile int _disposed = 0;
```

Comment:
We could set `_project` reference to `null` when disposing instead.

Reply:
But then _project wouldnt be readonly. :frowning: